libraries/AP_Compass/AP_Compass.cpp

```diff
@@ -502,20 +518,27 @@ void Compass::_detect_backends(void)
         return;
     }
 
+#if CONFIG_HAL_BOARD == HAL_BOARD_PX4
+    if (AP_BoardConfig::get_board_type() == AP_BoardConfig::PX4_BOARD_PIXHAWK2) {
+        // default to disabling LIS3MDL on pixhawk2 due to hardware issue
+        _driver_type_mask.set_default(1U<<DRIVER_LIS3MDL);
+    }
+#endif
+    
 /*
   macro to add a backend with check for too many backends or compass
   instances. We don't try to start more than the maximum allowed
  */
-#define ADD_BACKEND(backend, name, external) \
-    do { _add_backend(backend, name, external);     \
+#define ADD_BACKEND(driver_type, backend, name, external)   \
+    do { if (_driver_enabled(driver_type)) { _add_backend(backend, name, external); } \
        if (_backend_count == COMPASS_MAX_BACKEND || \
```

Comment:
hmm, but _backend_count is incremeneted after the check inside _add_backend()
So the return check is to ensure we don't try to load any more drivers if we already have 3. The check in _add_backend() is to make sure we don't already have 3 when we call it (which would have to be a bug, caused by someone not using ADD_BACKEND() macro)